File: 2020/01/advent202001.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	data, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	parsedInput := []int{}
	for _, s := range input {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		parsedInput = append(parsedInput, n)
	}
	fmt.Println(part1(parsedInput))
	fmt.Println(part2(parsedInput))
}

func part1(input []int) int {
	seen := make(map[int]bool)
	for _, n := range input {
		if seen[2020-n] {
			return n * (2020 - n)
		}
		seen[n] = true
	}
	return -1
}

func part2(input []int) int {
	sort.Sort(sort.IntSlice(input))
	z := len(input)
	for i := 0; i < z-2; i++ {
		j := i + 1
		k := z - 1
		for j < k {
			sum := input[i] + input[j] + input[k]
			if sum == 2020 {
				return input[i] * input[j] * input[k]
			}
			if sum < 2020 {
				j++
			} else {
				k--
			}
		}
	}
	return -1
}

File: 2020/01/advent202001.py
from typing import List

def part1(numbers: List[int], target: int) -> int:
  left, right = 0, len(numbers) - 1
  while numbers[left] + numbers[right] != target:
    if numbers[left] + numbers[right] > target:
      right -= 1
    else:
      left += 1
  return numbers[left] * numbers[right]

def part2(numbers: List[int], target: int) -> int:
  n = len(numbers)
  for i in range(n - 2):
    left, right = i + 1, n - 1
    while left < right:
      total = numbers[i] + numbers[left] + numbers[right]
      if total == target:
        return numbers[i] * numbers[left] * numbers[right]
      if total > target:
        right -= 1
      else:
        left += 1
  return -1

if __name__ == "__main__":
  with open('input.txt', 'r') as file:
    numbers = [int(num) for num in file]
  numbers.sort()
  target = 2020
  print(part1(numbers, target))
  print(part2(numbers, target))